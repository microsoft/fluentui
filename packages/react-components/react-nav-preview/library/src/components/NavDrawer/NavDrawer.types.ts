import { DrawerProps, DrawerSlots, DrawerState } from '@fluentui/react-drawer';
import { ComponentProps } from '@fluentui/react-utilities';
import { NavProps } from '../Nav/Nav.types';
import { NavContextValue } from '../NavContext.types';

/**
 * NavDrawer slots
 */
export type NavDrawerSlots = DrawerSlots;

/**
 * NavDrawer Props
 */
export type NavDrawerProps = ComponentProps<NavDrawerSlots> &
  DrawerProps &
  NavProps & {
    /**
     * The component uses arrow navigation by default.
     * Setting this to true enables tab AND arrow navigation.
     * @default false
     */
    tabbable?: boolean;
  };

/**
 * State used in rendering NavDrawer
 */
export type NavDrawerState = DrawerState & NavContextValue;
