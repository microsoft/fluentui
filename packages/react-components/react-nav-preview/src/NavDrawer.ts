export * from './components/NavDrawer/index';
