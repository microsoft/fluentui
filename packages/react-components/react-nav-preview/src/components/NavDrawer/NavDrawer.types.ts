import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';

export type NavDrawerSlots = {
  root: Slot<'div'>;
};

/**
 * NavDrawer Props
 */
export type NavDrawerProps = ComponentProps<NavDrawerSlots> & {};

/**
 * State used in rendering NavDrawer
 */
export type NavDrawerState = ComponentState<NavDrawerSlots>;
// TODO: Remove semicolon from previous line, uncomment next line, and provide union of props to pick from NavDrawerProps.
// & Required<Pick<NavDrawerProps, 'propName'>>
