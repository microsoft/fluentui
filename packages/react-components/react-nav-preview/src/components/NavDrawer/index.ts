export * from './NavDrawer';
export * from './NavDrawer.types';
export * from './renderNavDrawer';
export * from './useNavDrawer';
export * from './useNavDrawerStyles.styles';
