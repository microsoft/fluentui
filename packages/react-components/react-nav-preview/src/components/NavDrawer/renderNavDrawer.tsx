/** @jsxRuntime automatic */
/** @jsxImportSource @fluentui/react-jsx-runtime */

import { assertSlots } from '@fluentui/react-utilities';
import type { NavDrawerState, NavDrawerSlots } from './NavDrawer.types';

/**
 * Render the final JSX of NavDrawer
 */
export const renderNavDrawer_unstable = (state: NavDrawerState) => {
  assertSlots<NavDrawerSlots>(state);

  // TODO Add additional slots in the appropriate place
  return <state.root />;
};
