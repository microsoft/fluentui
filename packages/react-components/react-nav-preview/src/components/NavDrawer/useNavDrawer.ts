import * as React from 'react';
import { getIntrinsicElementProps, slot } from '@fluentui/react-utilities';
import type { NavDrawerProps, NavDrawerState } from './NavDrawer.types';

/**
 * Create the state required to render NavDrawer.
 *
 * The returned state can be modified with hooks such as useNavDrawerStyles_unstable,
 * before being passed to renderNavDrawer_unstable.
 *
 * @param props - props from this instance of NavDrawer
 * @param ref - reference to root HTMLDivElement of NavDrawer
 */
export const useNavDrawer_unstable = (props: NavDrawerProps, ref: React.Ref<HTMLDivElement>): NavDrawerState => {
  return {
    // TODO add appropriate props/defaults
    components: {
      // TODO add each slot's element type or component
      root: 'div',
    },
    // TODO add appropriate slots, for example:
    // mySlot: resolveShorthand(props.mySlot),
    root: slot.always(
      getIntrinsicElementProps('div', {
        ref,
        ...props,
      }),
      { elementType: 'div' },
    ),
  };
};
