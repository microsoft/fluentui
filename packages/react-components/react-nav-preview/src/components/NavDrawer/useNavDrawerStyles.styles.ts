import { makeStyles, mergeClasses } from '@griffel/react';
import type { SlotClassNames } from '@fluentui/react-utilities';
import type { NavDrawerSlots, NavDrawerState } from './NavDrawer.types';

export const navDrawerClassNames: SlotClassNames<NavDrawerSlots> = {
  root: 'fui-NavDrawer',
  // TODO: add class names for all slots on NavDrawerSlots.
  // Should be of the form `<slotName>: 'fui-NavDrawer__<slotName>`
};

/**
 * Styles for the root slot
 */
const useStyles = makeStyles({
  root: {
    // TODO Add default styles for the root element
  },

  // TODO add additional classes for different states and/or slots
});

/**
 * Apply styling to the NavDrawer slots based on the state
 */
export const useNavDrawerStyles_unstable = (state: NavDrawerState): NavDrawerState => {
  const styles = useStyles();
  state.root.className = mergeClasses(navDrawerClassNames.root, styles.root, state.root.className);

  // TODO Add class names to slots, for example:
  // state.mySlot.className = mergeClasses(styles.mySlot, state.mySlot.className);

  return state;
};
