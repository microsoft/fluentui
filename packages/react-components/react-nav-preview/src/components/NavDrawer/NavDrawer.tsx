import * as React from 'react';
import type { ForwardRefComponent } from '@fluentui/react-utilities';
// import { useCustomStyleHook_unstable } from '@fluentui/react-shared-contexts';
import { useNavDrawer_unstable } from './useNavDrawer';
import { renderNavDrawer_unstable } from './renderNavDrawer';
import { useNavDrawerStyles_unstable } from './useNavDrawerStyles.styles';
import type { NavDrawerProps } from './NavDrawer.types';

/**
 * NavDrawer component - TODO: add more docs
 */
export const NavDrawer: ForwardRefComponent<NavDrawerProps> = React.forwardRef((props, ref) => {
  const state = useNavDrawer_unstable(props, ref);

  useNavDrawerStyles_unstable(state);
  // TODO update types in packages/react-components/react-shared-contexts/src/CustomStyleHooksContext/CustomStyleHooksContext.ts
  // https://github.com/microsoft/fluentui/blob/master/rfcs/react-components/convergence/custom-styling.md
  // useCustomStyleHook_unstable('useNavDrawerStyles_unstable')(state);
  return renderNavDrawer_unstable(state);
});

NavDrawer.displayName = 'NavDrawer';
