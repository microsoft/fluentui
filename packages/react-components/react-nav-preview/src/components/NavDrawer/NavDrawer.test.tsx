import * as React from 'react';
import { render } from '@testing-library/react';
import { isConformant } from '../../testing/isConformant';
import { NavDrawer } from './NavDrawer';

describe('NavDrawer', () => {
  isConformant({
    Component: NavDrawer,
    displayName: 'NavDrawer',
  });

  // TODO add more tests here, and create visual regression tests in /apps/vr-tests

  it('renders a default state', () => {
    const result = render(<NavDrawer>Default NavDrawer</NavDrawer>);
    expect(result.container).toMatchSnapshot();
  });
});
