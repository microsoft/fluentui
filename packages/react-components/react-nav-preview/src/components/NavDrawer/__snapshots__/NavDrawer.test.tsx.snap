// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavDrawer renders a default state 1`] = `
<div>
  <div
    class="fui-NavDrawer"
  >
    Default NavDrawer
  </div>
</div>
`;
